Created Error page

sanketbansal opened the pull request:
<!--
(Thanks for sending a pull request! .)
-->

<!--The PR title should start with "Fix #bugnum: " (if applicable), followed by a clear one-line present-tense summary of the changes introduced in the PR. For example: "Fix #bugnum: Introduce the first version of the collection editor." -->

<!-- Add issue number here. If you do not solve the issue entirely, please change the message e.g. "Addresses #IssueNumber -->


#### Changes proposed in this pull request:

<!-- Changes: Add here what changes were made in this issue and if possible provide links. -->
- Added html-error component which is a parent component for error pages.
- Added server-error component for error 500.
- Refactor not-found component inside the parent component.
- Added routing for error pages.

<!-- Demo Link: Add here the link where you changes can be seen. -->
- Link to live demo: http://pr-122-evalai.surge.sh <!-- Replace XXX with your PR no: -->

<!-- Screenshots for the change: Add here the screenshot of the fix. -->
- 
![error](https://user-images.githubusercontent.com/17106489/55672258-a1999580-58b6-11e9-9fc7-c007e15821b0.gif)

Shekharrajak committed:
Initial error-pages/internal_server component created

Shekharrajak committed:
500 internal server error page will be redirected using http interceptor

Shekharrajak committed:
specific version for NgxTwitterTimelineModule

Shekharrajak committed:
error handler service is added

Shekharrajak committed:
lint error fixed

Shekharrajak committed:
typescript '>=2.4.2 <2.7.0' should be used for current Angular CLI and compiler version

Shekharrajak committed:
trying differenct version of typescript to run ng test command in travis

RishabhJain2018 committed:
Merge branch 'master' into gsoc_idea

sanyamdogra committed:
Fix #72: Add Code of Conduct(#93)

sanketbansal committed:
Created Error page

Added html-error component, server-error component.
Refactor server-error and not-found component as child of html-error component.
Updated app.module.ts and app-routing.module.ts

sanketbansal committed:
Developed server-error page.

Added error 500 gif in the html and the go back link.
changed CSS properties.

sanketbansal committed:
Fixed import bug

Import path in not-found.scss has been fixed.

lunayach commented:
@sanketbansal could you please check, why the Travis-build is failing?

sanketbansal commented:
I checked the travis-build issue. It is giving errors for ng test command.

### ng test --watch=false --progress=false --code-coverage > test-log.
<img width="615" alt="Capture" src="https://user-images.githubusercontent.com/17106489/55679635-7e113200-592c-11e9-9326-3f1b2223d9c1.PNG">

Any idea??

Shekharrajak commented:
@sanketbansal ,structure looks good, but :

- I didn't see any Interceptor concept in this PR.
- why did you choose html-error name ? Isn't `error-page` good name?
- How it will redirect to 500 or 404 page, I didn't find it (maybe I am missing something)

Shekharrajak reviewed (commented):
@sanketbansal ,structure looks good, but :

- I didn't see any Interceptor concept in this PR. Didn't you find it useful in https://github.com/Cloud-CV/EvalAI-ngx/pull/104 ? You can cherry-pick or add your commits top of my branch.
- why did you choose html-error name ? Isn't `error-page` good name?
- How it will redirect to 500 or 404 page, I didn't find it (maybe I am missing something)

sanketbansal committed:
Fixed testing issue.

Added Server-Error component and Not-Found component in the html-errors component spec.ts file.

sanketbansal commented:
@Shekharrajak.
- I created the PR as it has a different directory structure for error pages and new UI for error 500 component.
- There is no specific reason for naming to html-errors instead of error-page. U want me to update it?
- It redirects to errors for route "error/404" and "error/500".

Shekharrajak commented:
Thanks for letting me know these points.


> * I created the PR as it has a different directory structure for error pages and new UI for error 500 component.
> * There is no specific reason for naming to html-errors instead of error-page. U want me to update it?

yes, that I saw. I see that AngularJS project have directory name `error-page`, so it will be better to make it consistent. 

> * It redirects to errors for route "error/404" and "error/500".

Why not you start working on adding logic for it. It should be functional when backend server crashes/sending 500 status code/not found status for the request.

sanketbansal committed:
Added Integration Testing

Integration testing added for server-error component.
It checks the value of  <a> to be " EvalAi".
Also, <a> attribute 'href' value should be ''.

wadleo committed:
Fix #79: Compress images used in get involved page(#95)

sanketbansal committed:
Fix #108: Removes unnecessary horizontal & vertical scroll-bars from the home page(#103)

Removed changes from app.module.ts and package.json

sanketbansal committed:
UI fixes for login & sign up button on small screens(#113)

RishabhJain2018 commented:
@lunayach @Shekharrajak Can you guys please review this PR and get it merged.

Commit:
Update README.md with backend installation instructions(#115)

Shekharrajak commented:
@sanketbansal let us know the status of the PR.

Commit:
fix: Fix Endpoint Service not being injected in self object (#117)

* Endpoint Service was not being injected in self object

* Add Endpoint service in test file providers

RishabhJain2018 committed:
Readme: Add new mentors in team section(#128)

sanketbansal commented:
It is in progress I need to commit some changes.

RishabhJain2018 committed:
Environment: Add base API endpoint URLs for local and staging(#127)

sanketbansal committed:
Renamed html-errors

Renamed the html-errors folder to errors

sanketbansal committed:
Added logic to navigate to error pages

Routing added in handleApiError method to navigate to different error pages

sanketbansal committed:
Added RouterTestingModule

RouterTestingModule include in spec.ts file to resolve testing errors.

RishabhJain2018 commented:
@sanketbansal Any updates on this?

sanketbansal committed:
Navigate to server-error page

Changed logic to navigate to server-error page

sanketbansal commented:
@RishabhJain2018  please review the PR

RishabhJain2018 commented:
@Shekharrajak @lunayach Can you please review this as well?

RishabhJain2018 committed:
Setup: Fix docker configuration to set up evalai-ngx in dev environment(#129)

Shekharrajak commented:
I don't see reply for my review comments above so I am not reviewing it again and assuming it was not read.

sanketbansal commented:
@Shekharrajak  I have completed the changes in the PR.
Tell me if there is some other changes required to be done.

Shekharrajak commented on src/app/components/challenge/challengeevaluation/challengeevaluation.component.spec.ts:
```diff
@@ -16,7 +17,7 @@ describe('ChallengeevaluationComponent', () => {
       declarations: [ ChallengeevaluationComponent ],
       providers: [ ChallengeService, ApiService, GlobalService, AuthService, EndpointsService ],
-      imports: [ HttpClientModule ]
+      imports: [ HttpClientModule, RouterTestingModule ]
```
I don't see this change relevant to this PR.

sanketbansal replied:
ok i will drop that change

Shekharrajak commented on src/app/components/challenge/challengeoverview/challengeoverview.component.spec.ts:
```diff
@@ -16,7 +17,7 @@ describe('ChallengeoverviewComponent', () => {
       declarations: [ ChallengeoverviewComponent ],
       providers: [ ChallengeService, ApiService, GlobalService, AuthService, EndpointsService ],
-      imports: [ HttpClientModule ]
+      imports: [ HttpClientModule, RouterTestingModule ]
```
I don't see this change relevant to this PR.

sanketbansal replied:
ok i will drop that change

Shekharrajak replied:
The thing is if you have mentioned that PR is complete that means you are 100% sure that you have reviewed and tested your changes properly. Iterating the PR review multiple times is not a good thing.

sanketbansal replied:
@Shekharrajak  i checked this change again actually it is required for testing else travis CI will give testing error. So will have to keep it.

Shekharrajak replied:
Thanks for letting me know this point. You can paste the error here (or link of travis error line), so that it can be tracked later.

Shekharrajak commented on src/app/components/errors/server-error/server-error.component.spec.ts:
```diff
@@ -0,0 +1,40 @@
+    expect(de.query(By.css('p')).query(By.css('a')).nativeElement.innerText).toBe(' EvalAi');
+  });
+
+  it('should be redirected to "" ', () => {
```
' .. redirected to homepage..' is readable

sanketbansal replied:
okay will change that

Shekharrajak commented on src/app/components/errors/server-error/server-error.component.html:
```diff
@@ -0,0 +1,3 @@
+
+<img src="../../../../assets/images/500-Internal-Error.gif">
```
Should we use our own created GIF image here or downloaded image is fine as well?

sanketbansal replied:
The created GIF will be good but as long as not much customization is required we can use downloaded images as well.

Shekharrajak replied:
@RishabhJain2018 , do EvalAI AngularJS uses own created images or downloaded one? Should we use EvalAI images/SVG only ? @lunayach

Shekharrajak commented on src/app/services/global.service.ts:
```diff
@@ -321,8 +322,12 @@ export class GlobalService {
       this.showToast('error', err.error['error'], 5);
```
Can we use this service and this ` handleApiError`  instead  [this service](https://github.com/Cloud-CV/EvalAI-ngx/pull/104/files#diff-477bb86affda2e5cdadf1244bbe6757f) ?

sanketbansal replied:
If u mean whether we should this handleApiError or [this service](https://github.com/Cloud-CV/EvalAI-ngx/pull/104/files#diff-477bb86affda2e5cdadf1244bbe6757f).
then this handleApiError method would be better as it handles all the api errors. Therefore adding the logic here itself might be correct.

Shekharrajak reviewed (commented):
I don't see reply for https://github.com/Cloud-CV/EvalAI-ngx/pull/122#pullrequestreview-223587455 comment and commits from PR #104

sanketbansal commented:
> I don't see a reply for [#122 (review)](https://github.com/Cloud-CV/EvalAI-ngx/pull/122#pullrequestreview-223587455) comment and commits from PR #104

@Shekharrajak I did not **cherrypicked** the commit of #104 because this PR has already similar commits present.
Tell me if it is required. I will do that.

Shekharrajak commented:
Cherry picking means - You are reserving the original author and authentication of commit (rather than copying the code and using it).

> @Shekharrajak I did not **cherrypicked** the commit of #104 because this PR has already similar commits present.

I don't see Interceptor logic in this PR. Please let me know what are the things you see similar (except [this](https://github.com/Cloud-CV/EvalAI-ngx/pull/122#discussion_r284730412))

sanketbansal committed:
Resolved merge-conflict build errors

Added FeaturedChallenge in app.module.ts

sanketbansal committed:
Fixed testing error

Added RouterTestingModule

sanketbansal committed:
Changed typescript version

Typescript version changes to "^2.5.3"

sanketbansal committed:
Renamed test case heading

Test case renamed to 'should be redirected to homePage'

sanketbansal commented:
@Shekharrajak,  I have rebased this PR with #104. Now, should I drop some similar changes present after rebasing like internal server error page?

sanketbansal committed:
Dropped internalServerError component

Dropped internal server error component and deleted the dependencies over it.

sanketbansal committed:
changed routing

changed routing to 'error/500'

Shekharrajak commented on CODE_OF_CONDUCT.md:
```diff
@@ -0,0 +1,44 @@
+# Contributor Code of Conduct
```
Not sure why this changes came here.

Shekharrajak commented on src/app/interceptors/http-error-interceptor.ts:
```diff
@@ -0,0 +1,30 @@
+
+    return next.handle(request).do((event: HttpEvent<any>) => { }, (err: any) => {
+      if (err instanceof HttpErrorResponse) {
+        this.errorHandler.handleError(err);
```
Here we can use already present service - `services/global.service.ts` - ` handleApiError method`

sanketbansal replied:
okay will do that.

Shekharrajak commented on src/app/services/error-handler.service.ts:
```diff
@@ -0,0 +1,23 @@
+import { Injectable } from '@angular/core';
```
this service can be removed

sanketbansal replied:
okay

Shekharrajak commented:
Yes! now you can do your changes top of it. Please have a look, into the comments

gautamjajoo committed:
Add CONTRIBUTING.md file (#112)

* Create CONTRIBUTING.md

* Add Gitter link

* Add ng-test

* Update CONTRIBUTING.md

sanketbansal committed:
Merge remote-tracking branch 'origin/Html-errors' into Html-errors

sanketbansal committed:
Removed error handler service

Error handler service removed from the application

sanketbansal committed:
Added global service for interceptor

GlobalService provided for the interceptor to handleAPIError.
